### PR TITLE
feat: Adds favorite departures (closes #680)

### LINKED ISSUE #680: [Feature] Add favorite departures
## Origin
_Links to received feedback, user research or other findings._

See issue #420 

## Motivation
_A short description of what user needs or business goals this feature will solve._

Allow the user to save a frequently used departure.

## Hypotheses and assumptions
_A list of hypotheses and assumptions we have made about the user or the proposed solution._

Many users use the same line every day from the same stop, and want to easily find these departures.

## Proposed solution
_A coarse description of a proposed solution, that may include wireframes and graphic design._

## Acceptance Criteria
- [ ] Able to add favorite.
- [ ] Favorites are persisted after app restart.
- [ ] I can remove favorite by toggling star.
- [ ] Able to filter on favorites nearby, by address search and from different quays.
- [ ] Able to add favorites from all types of transportation modes and quays.
- [ ] As a user of screen reader I can navigate and easily understand favorites and how to toggle between.
- [ ] Works as expected in dark mode.
- [ ] (Is translatable to English)
- [ ] Does not drain battery
- [ ] Works in up to 200% font size

Adding:
- "Star" from departures view
image

Using:
- Filter toggle on departures screen
- Toggle only shown if current results contain a favorite
image

Figma / App 1.0 / Favorite https://www.figma.com/file/WsvD8b5PnUwvvRJiyRe6I9/App-1.0?node-id=8%3A1529
Figma / App 1.0 / Departure https://www.figma.com/file/WsvD8b5PnUwvvRJiyRe6I9/App-1.0?node-id=4%3A982

### PR DESCRIPTION
**Pull request by mikaelbr**
**This is dependent on production deployment of the BFF. It is also dependent on #775 which should be merged first**

This adds favorite departures from the departure screen. Only scoped to #680 where we can add favorites directly from results. Currently, this is the only way to add favorite departures.

For Code Review, see Acceptance Criteria on #680.

## Examples

image

image

### CODE REVIEW

**Comment by rikkekantega** [comment]
Seems like using the toggler back and forth to fast (before result has loaded) triggers an error
image

I also managed to get an error when removing favorite. Trying to reproduce to describe steps.

**Comment by rikkekantega** [comment]
Okay, so I'm triggering quite a few errors when clicking around. I'm trying to find consistent reproduction steps. But I expect you would be able to trigger some by combining toggling favorite and adding/removing favorites back and forth.

**Comment by mikaelbr** [comment]
Hm. Maybe there are some regressions after my git management also 😬  I'm checking it out

File: src/screens/Nearby/NearbyResults.tsx [hunk 1]
```diff
@@ -1,93 +1,69 @@
...
+              ? t(NearbyTexts.results.messages.initial)
+              : t(NearbyTexts.results.messages.initial)}
+          </ThemeText>
         </MessageBox>
       </View>
     );
   }
 
-  if (hasNoQuays(departures)) {
+  if (hasNoQuaysWithDepartures(departures)) {
```
Comment (rikkekantega) [inline src/screens/Nearby/NearbyResults.tsx:53]: Wonder if loading state should be considered here?

Comment (mikaelbr) [reply src/screens/Nearby/NearbyResults.tsx:53]: Sure!

File: src/components/sections/action-item.tsx [hunk 2]
```diff
@@ -13,20 +13,22 @@ export type ActionItemProps = SectionItem<{
...
+  accessibility,
   ...props
 }: ActionItemProps) {
   const {contentContainer, topContainer} = useSectionItem(props);
   const style = useSectionStyle();
 
   if (mode === 'toggle') {
     return (
-      <View style={[style.baseItem, style.spaceBetween, topContainer]}>
+      <View style={[style.spaceBetween, topContainer]} {...accessibility}>
```
Comment (rikkekantega) [inline src/components/sections/action-item.tsx:31]: Generally, corresponding label should respond to click action by toggling the switch, 
and I wonder if we will make the function more understandable by doing so as well. If the entire containing element has role switch, and an onPress action. 
Right now, for the usage in favorites toggler, talkback reads "Vis favoritter" (no action provided), and then by navigating to the next item, the switch is read out so fast that I cant interpret what is said.  (only the following hint about how to interact, which I think may be Android specific?) Probably default behaviour, but I'm not sure if it's understandable enough in our case

File: src/translations/screens/Nearby.ts [hunk 3]
```diff
@@ -39,20 +42,63 @@ const NearbyTexts = {
...
+    lines: {
+      a11y: {
+        line: _(`Linje:`),
+      },
+      lineNameAccessibilityHint: _(
+        'Aktiver for detaljer om avgang og oversikt over kommende avganger.',
+      ),
+      favorite: {
+        addFavorite: (name: string, place: string) =>
+          _(`Legg til favorittavgang: ${name} fra ${place}`),
```
Comment (rikkekantega) [inline src/translations/screens/Nearby.ts:83]: Really nice!!! :raised_hands:

File: src/components/sections/action-item.tsx (hunk 2 above)
Comment (mikaelbr) [reply src/components/sections/action-item.tsx]: > Generally, corresponding label should respond to click action by toggling the switch,

Maybe there are different practices on Android / iOS here? I know checkboxes and labels are a thing on the Web but typically iOS pattern is to _not_ have the label (entire line) be pressable to toggle Switches.

Comment (mikaelbr) [reply src/components/sections/action-item.tsx]: Should we possibly do text as inaccessible and adding an accessible label equal to the text on the Switch itself?

**Comment by mikaelbr** [comment]
Errors you (@rikkekantega ) saw was caused by layout animations 😢  Deactivated on Android for now. Too bad, but can look into how we can get it to work later.

File: src/components/sections/action-item.tsx (hunk 2 above)
Comment (rikkekantega) [reply src/components/sections/action-item.tsx]: Okay, seems like clickable label for switch is an Android-only pattern then.

Comment (rikkekantega) [reply src/components/sections/action-item.tsx]: Your suggestion on labeling the switch: Yes, I think that would make more sense =)

Comment (mikaelbr) [reply src/components/sections/action-item.tsx]: Hm. Then we should probably do so on android. I can look into it

Comment (mikaelbr) [reply src/components/sections/action-item.tsx]: I did some accessibility improvements. But not the clickable thing. I think we should work out a good way to do both patterns on the platforms, but tempted to say we should do that as a separate issue as it affects many other things aswell

**Comment by rikkekantega** [comment]
image
Label/star will at least need some space between to be accessible for people with reduced precision (and to people with clubbed thumbs, like me :grimacing: )

**Comment by mikaelbr** [comment]
Have you checked the latest version, @rikkekantega ? I bumped the hitslop some commits ago

**Comment by rikkekantega** [comment]
> Have you checked the latest version, @rikkekantega ? I bumped the hitslop some commits ago

Yep, I just recently pulled, hitslop is set to 14, 8

**Comment by mikaelbr** [comment]
Added some additional spacing.

**Comment by rikkekantega** [comment]
200%
Long line names push favorite icon out of view
image

Switch is hard to hit because it's pushed or too close to text. Label text should probably wrap sooner / have some more margin
image

Quay name should also wrap to avoid colliding/pushing distance
image

**Comment by mikaelbr** [comment]
Good point! Lets do it as a part of the qa flow

**Pull request merged**